src/calling/haplotypes.rs

```diff
@@ -166,3 +172,54 @@ impl KallistoEstimates {
         Ok(filtered)
     }
 }
+
+#[derive(Derefable, Debug, Clone, PartialEq, Eq, Hash, new)]
+pub(crate) struct Variant(#[deref] String);
+
+#[derive(Derefable, Debug, Clone, PartialEq, Eq)]
+pub(crate) struct HaplotypeVariants(#[deref] HashMap<Variant, Vec<Haplotype>>);
+
+impl HaplotypeVariants {
+    pub(crate) fn new(haplotype_variants: &mut bcf::Reader) -> Result<Vec<HaplotypeVariants>> {
+        let mut vec_haplotype_variants = Vec::new();
+        for record_result in haplotype_variants.records() {
+            let record = record_result.expect("Failed to read record");
+
+            //store the variant
+            let mut variant = String::new();
+            variant.push_str(&format!("{}{}", record.contig(), "_"));
+            let pos = record.pos().to_string();
+            variant.push_str(&format!("{}{}", pos, "_"));
+            for allele in record.alleles() {
+                for c in allele {
+                    let base = char::from(*c).to_string();
+                    variant.push_str(&format!("{}{}", base, "_"));
+                }
+            }
+
+            //store the haplotypes that carry the variant
+            let header = record.header();
+            let gts = record.genotypes().expect("Error reading genotypes"); //genotypes of all samples
+            let sample_count = usize::try_from(record.sample_count()).unwrap();
+            let sample = (0..sample_count).collect::<Vec<_>>();
+            let mut haplotypes: Vec<Haplotype> = Vec::new();
+
+            for (sample_index, mut x) in sample.iter().zip(header.samples().into_iter()) {
```

Comment:
To get the index of each sample/haplotype, just use `enumerate()` on the iterator over the `samples()`.